### PR TITLE
add maintainer scripts for haskell package generation

### PR DESCRIPTION
**Pull request by maralorn**
This is a slightly reworked version of #86699. Thx again @hyperfekt.

Introduces a script that can be used to update the Nix expressions for
the Haskell package set. In service of that, also
- introduces cabal2nix-latest, which pins the hackage2nix version used
- changes all-cabal-hashes to use fetchFromGitHub
- adds update-hackage.sh & update-cabal2nix-latest.sh maintainer scripts

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by maralorn** [comment]
Okay, as I feared this PR breaks the callHackage function, because it expects all-cabal-hashes to be a tarball.

**Comment by maralorn** [comment]
There are two ways out. Either change callHackage to not expect a tarball or change this PR to not unpack.

I think the second is the way to go, because we are talking about _a lot_ of files and directories here and it seems quite inefficient. But I may be unaware of other arguments or it might be that we pay the performance cost anyways …

**Comment by hyperfekt** [comment]
I've packaged ratarmount and am currently reworking my PR to use the compressed version of `all-cabal-hashes` instead. Ideally cabal2nix would adopt the same technique cabal does to work directly with the compressed archive (as suggested by peti), but in lieu of that it seems like a workable solution.

**Comment by maralorn** [comment]
@hyperfekt Oh, I have already pushed a fixed version of this PR. What do you think?

**Comment by hyperfekt** [comment]
I think that the last time I tried to unpack that archive on ext4 it took dozens of minutes if I recall correctly. ^^
Not to mention that it's 700MB big, which even considering GCing is a *lot*.

**Comment by maralorn** [comment]
Yes, put since hackage2nix expects it to be unpacked we don‘t have a choice, do we?

At least this way people using callHackage don‘t need to unpack it. Only people who run the regenerate script.

**Comment by maralorn** [comment]
But, yeah, if the ratarmount solution is cooler. Let’s go for it.

**Comment by hyperfekt** [comment]
That is why I packaged ratarmount. It allows us to mount the archive without actually unpacking it onto the disk. I'll see in a bit how fast that ends up being.

**Comment by maralorn** [comment]
Yeah, go for it!

**Comment by maralorn** [comment]
Sadly there might be a trade-off. Unpacking with nix once is probably slow on the first run, but maybe it‘ll be faster when rerunning?

**Comment by hyperfekt** [comment]
Ideally it'll be almost as fast, at least if I can get Hydra to build the index file for the archive once and have that be used by everyone for every run.

**Comment by sternenseemann** [comment]
I would rather rework `callHackage` and have `all-cabal-hashes` be the unpacked version. Having a compressed store path has no real advantages in that case because we compress them anyways while fetching from a binary cache. However mounting a tar archive will probably have significant impact on either RAM usage or access performance over having the archive already unpacked in store.

**Comment by maralorn** [comment]
I don‘t have a strong opinion about this. (And the old solution still in the reflog.)

But it’s a 700MB file that everyone would need to write to disk and keep around. I am not exactly sure what callHackage does with the tarball but it seems pretty fast. So when we can get hackage2nix to also use that tarball with reasonable speed I‘d prefer that.

I personally can spare the RAM more than the disk space … For a real principled decision we would need measurements. But that’s likely not worth the effort.

**Comment by hyperfekt** [comment]
The only reason I got into this in the first place was that the number of files was so high that some file systems had serious trouble with it.
But after measuring it the ratarmount version is almost 5 times slower at regenerating hackage-packages.nix.
I guess that slowness for some is just something we'll have to live with, rather than making it this much slower for everyone.
Leaving all-cabal-hashes uncompressed would mean that anyone using callHackage gets 700MB dumped into their store (filesystem compression is presumably without effect because these are tiny files) , versus only the people regenerating the Haskell package set.

**Comment by sternenseemann** [comment]
`callHackage` extracts only the necessary subset of the `tar`: <https://github.com/NixOS/nixpkgs/blob/d05202ea7c026640c18c9d058aa24a0469e7930d/pkgs/development/haskell-modules/make-package-set.nix#L144-L148>

I guess ratarmount is fine for now, we can always change things if we run into problems.

File: maintainers/scripts/haskell/regenerate-hackage-packages.sh
Comment (hyperfekt) [inline maintainers/scripts/haskell/regenerate-hackage-packages.sh:17]: This should unpack directly to $out instead of using the unpackPhase, since it takes almost twice as long otherwise due to having to move all files into the store.

**Comment by maralorn** [comment]
@sternenseemann Well, this PR goes the middle ground. callHackage stays efficient but for actually doing the update it will extract into the nix-store. (Which apparently is faster than using ratarmount.)

File: maintainers/scripts/haskell/regenerate-hackage-packages.sh
Comment (maralorn) [reply maintainers/scripts/haskell/regenerate-hackage-packages.sh:17]: Good catch.

Comment (maralorn) [reply maintainers/scripts/haskell/regenerate-hackage-packages.sh]: @hyperfekt Does the change I pushed look better?

File: maintainers/scripts/haskell/regenerate-hackage-packages.sh
Comment (hyperfekt) [inline maintainers/scripts/haskell/regenerate-hackage-packages.sh:20]: Ideally a comment would either indicate that an alternative hackage2nix can be passed to this script (which may be useful for developing fixes to hackage2nix), or the functionality would be removed; since the syntax for it isn't very obvious, well known, or googleable and will probably just become arcana without it.

File: maintainers/scripts/haskell/regenerate-hackage-packages.sh
Comment (hyperfekt) [reply maintainers/scripts/haskell/regenerate-hackage-packages.sh]: Yes, that should work well.

File: maintainers/scripts/haskell/regenerate-hackage-packages.sh
Comment (maralorn) [reply maintainers/scripts/haskell/regenerate-hackage-packages.sh:20]: Well then let’s remove it until someone needs it. It’s easy enough to add again.

File: pkgs/data/misc/hackage/default.nix
Comment (sternenseemann) [inline pkgs/data/misc/hackage/default.nix:7]: Can we add the update script via `passthru` here?

File: pkgs/development/haskell-modules/cabal2nix-latest.nix
Comment (sternenseemann) [inline pkgs/development/haskell-modules/cabal2nix-latest.nix:40]: Adding the update script here as well would be nice, but should probably be added via an override in `non-hackage-packages.nix` to avoid a terrible sed hack.

File: pkgs/data/misc/hackage/default.nix
Comment (maralorn) [reply pkgs/data/misc/hackage/default.nix:7]: Can you enlighten me as to who or what does actually use the passthru update script?

I am a bit sceptical because at least fetchgit didn‘t accept a passthru attribute.

Comment (sternenseemann) [reply pkgs/data/misc/hackage/default.nix:7]: `fetchurl` accepts `passthru`. IIRC some automation uses it.

File: pkgs/development/haskell-modules/cabal2nix-latest.nix
Comment (maralorn) [reply pkgs/development/haskell-modules/cabal2nix-latest.nix:40]: Okay, well. Now that I learned that I can use `nix-shell maintainers/scripts/update.nix --argstr package haskellPackages.cabal2nix-latest` I wonder if it’s a smart id to have the update script at such a prominent position as `maintainers/scripts/haskell`.

File: pkgs/data/misc/hackage/default.nix
Comment (maralorn) [reply pkgs/data/misc/hackage/default.nix]: Okay, done.

`nix-shell maintainers/scripts/update.nix --argstr package all-cabal-hashes` is an alias for our script now.^^

**Comment by teto** [comment]
The faith is reborn https://github.com/NixOS/nixpkgs/pull/62105#issuecomment-687763131

File: maintainers/scripts/haskell/update-stackage.sh
Comment (cdepillabout) [inline maintainers/scripts/haskell/update-stackage.sh:41]: Now that we're on Stackage Nightly, I imagine we can drop quite a few of these lines here, since these packages will generally also be the latest version on Hackage.  That is to say, the version of the package in Stackage Nightly should also be the latest version of the package on Hackage.

If the packages in Stackage Nightly are _not_ the latest version on Hackage, then the Stackage maintainers have probably done that for a good reason.  For instance, the packages aren't compatible and don't compile together.

Although this is not something that should block this PR.  We can do this at some point in  the future.

Comment (sternenseemann) [reply maintainers/scripts/haskell/update-stackage.sh:41]: We should drop every one of these except the ones that are nixpkgs-related `jailbreak-cabal`, `cabal2nix`, ...

**Review by cdepillabout — APPROVED** [review]
I tried out all of the maintainer scripts in this PR and they all seem to work well.

This LGTM.

File: maintainers/scripts/haskell/update-stackage.sh
Comment (maralorn) [reply maintainers/scripts/haskell/update-stackage.sh:41]: Yeah, let’s clean that up. But that requires a bit of careful deliberation to see if we break anything right now, so I would like this to happen in an independent PR.

**Comment by maralorn** [comment]
Thanks everyone! Especially to @teto and @hyperfekt for your tries of doing this.

**Comment by hyperfekt** [comment]
Thanks a lot to you as well! I really appreciate you taking this over the finish line; and I'm excited for the future of the nixpkgs Haskell package set.

**Pull request merged**
